Update: Allow string severity in config (fixes #3626)

nzakas opened the pull request:
It turns out this way pretty easy now that we have configs operations nicely split out. You can now do this:

```
some-rule: "error"
some-rule: ["warn", {}]
```

Also works with inline JSON comments.

**TODO:** Documentation

Questions:

* To what extent do we change the documentation?
* Do we just say to always use the string values?
* Do we say you can use either strings or numbers?
* How much rule documentation do we change?

nzakas commented on packages/eslint-config-eslint/default.yml:
```diff
@@ -2,93 +2,93 @@ extends:
-    no-alert: 2
-    no-array-constructor: 2
-    no-caller: 2
+    array-callback-return: "error"
```
:dog: :hamburger: (dogfood)

nzakas commented on tests/lib/config/config-ops.js:
```diff
@@ -297,7 +297,7 @@ describe("ConfigOps", function() {
 
             var config = [
-                { rules: { "no-mixed-requires": [1, "nconf", "underscore"] } },
+                { rules: { "no-mixed-requires": ["warn", "nconf", "underscore"] } },
```
Intermixed a few string severities in other tests for sanity checks.

platinumazure commented:
Nice!

Actually, that made me think about something else. Could we have a top-level config item where a user could specify other strings to be accepted, which we would read and use? An example being poor man's localization:

```json
{
    "ruleSeverities": {
        "error": "yarrrrrrrrrrr",
        "warn": "yar",
        "off": "nay"
    },
    "rules": {
        "semi": ["yarrrrrrrrrrr", "always"],
        "no-semi-spacing": "yar",
        "indent": "nay"
    }
}
```

platinumazure commented on tests/lib/config/config-ops.js:
```diff
@@ -481,4 +481,151 @@ describe("ConfigOps", function() {
+            var config = {
+                rules: {
+                    foo: ["error", "something"],
+                    bar: "error"
```
This line seems redundant (covered in above test) and also unnecessary cognitive load given the title of the test. Do you really need 2 rules in each test, anyway?

xjamundx replied:
Might be good to test more complex configs. Do all the various combinations works.

nzakas replied:
I have two rules because I'm always afraid of side effects that happens with multiple input types.

There are only two things to test: rules with arrays and rules without arrays. 

Remember, all this method does is translate string values into numeric values, it's not very complicated.

gyandeeps commented:
> Do we just say to always use the string values?

No

> Do we say you can use either strings or numbers?

Yes

@nzakas I think the current implementation is case-sensitive, we need to make it case-insensitive as per the discussion inside issue.

xjamundx commented on packages/eslint-config-eslint/default.yml:
```diff
@@ -2,93 +2,93 @@ extends:
+    indent: ["error", 4, {SwitchCase: 1}]
+    block-spacing: "error"
+    brace-style: ["error", "1tbs"]
+    camelcase: ["error", { properties: "never" }]
```
That really looks nice, especially in the `.yml` files which I think are meant to be fairly readable. This is great.

xjamundx commented:
Looks great :+1: 

I think this could be something to advertise and possibly to include in the `--init` stuff and document. I'd suggest it should be the preferred way and eventually considering deprecation for the numbers. This seems much more readable and it's not like saving bytes in config files is a huge burden. Seems great!

nzakas commented:
> I think the current implementation is case-sensitive, we need to make it case-insensitive as per the discussion inside issue.

@gyandeeps good point. I'll make that change.

nzakas commented:
Okay, carried this all the way through:

1. Updated recommended config to use strings
1. Updated config initializer to use strings
1. Added some more helper functions for easier translation between numbers and strings
1. Updated doc validation to look for "error" in recommended rules
1. Updated autoconfig to use strings
1. Updated configuration documentation

nzakas commented:
@eslint/eslint-team any feedback on this?

ilyavolodin commented on docs/user-guide/configuring.md:
```diff
@@ -318,21 +326,21 @@ And in YAML:
+  - plugin1
+rules:
+  eqeqeq: 0
+  curly: error
```
Should those be quoted? They are string values, right?

BYK replied:
We don't need quote these unless they have special characters in them.

ilyavolodin commented:
LGTM. Not sure about yaml quotes, not familiar enough with yaml to know if they are required.

BYK commented on lib/config/autoconfig.js:
```diff
@@ -312,8 +313,11 @@ Registry.prototype = {
+
     var recRules = Object.keys(recConfig.rules).filter(function(ruleId) {
-        return (recConfig.rules[ruleId] === 2 || recConfig.rules[ruleId][0] === 2);
+        return ConfigOps.isErrorSeverity(recConfig.rules[ruleId]);
```
I may have preferred normalizing to numbers and having a `string -> number` mapping somewhere but I guess this is also okay.

BYK commented on lib/config/config-ops.js:
```diff
@@ -21,6 +21,14 @@ var lodash = require("lodash"),
+    "off": 0
+};
+
+var RULE_SEVERITY_STRINGS = ["off", "warn", "error"];
```
Why not use `Object.keys(RULE_SEVERITY);`?

BYK replied:
Or if we're concerned about the order, the other way around:

```js
var RULE_SEVERITY_STRINGS = ["off", "warn", "error"];
var RULE_SEVERITY = RULE_SEVERITY_STRINGS.reduce(function (map, str, num) {
    map[str] = num;
    return map;
}, {});
```

Reasoning: avoiding repeating same values in two different places so it is less error-prone in case these values change.

nzakas replied:
Good point, I'll update that.

nzakas commented:
Updated. Last call for comments.
